README

  Welcome to Tile World

Tile World is an emulation of the game "Chip's Challenge" for the Atari
Lynx, created by Chuck Sommerville, and later ported to MS Windows by
Microsoft (among other ports).

  Important Note

Tile World is an emulation of the "Chip's Challenge" game engines only. It
does not come with the chips.dat file that contains the original level set.
That file, which is copyrighted and cannot be freely distributed, was
originally distributed with the MS version of "Chip's Challenge". If you
have a copy of this version of the game, you can use that file to play the
original games in Tile World. If you do not have a copy of this file,
however, you can still play Tile World with the many freely available level
files created by fans of the original game.

  Installing Tile World under Windows

First of all, you'll want to store the files contained in this archive into
its own separate directory. If you're using the self-extracting executable,
you can create a new directory during the installation. Otherwise, you'll
need to create a new directory beforehand -- something like c:\tworld --
and extract the files in there.

If you have a copy of the chips.dat data file, copy it to the data
subdirectory. This will allow you to play the original levels under Tile
World (for the MS ruleset and the Lynx ruleset both).

If you have other data files that you would like to try out in Tile World,
copy those to the sets directory.

The shell commands to do the above would look something like:

  cd c:\wherever\my\copy\of\chips\challenge\is\at
  copy chips.dat c:\tworld\data
  cd c:\my\collection\of\dat\files
  copy *.dat c:\tworld\sets

That's all that needs to be done to set it up. Run the program as
c:\tworld\tworld2, or create a shortcut for it.

  Installing Tile World under Linux (or building from source on Windows)

Tile World 2 has not yet been extensively tested on Linux.

The build system at this point is not extremely robust, consisting of
only a hand-written makefile for GNU Make and no configure script.

(Efforts are underway to improve upon this, and to possibly also provide
pre-compiled binary packages for some flavors of Linux.)

Before proceeding, ensure that you have Qt 4 and SDL installed on your machine.
(If you don't have SDL, you can get it by visiting
http://www.libsdl.org/download.html. If you download a precompiled version
-- i.e., an .rpm or .deb file -- note that you will need the development
runtime, as opposed to the binary runtime.)
Tile World has been tested with Qt 4.8. Earlier versions of Qt may not work.

Installing Tile World involves the following steps:

Assuming that the Qt and SDL development libraries are properly installed,
the makefile will try to figure out the necessary flags internally by calling
qmake and sdl-config.

If the environment variable QTDIR is set, then $QTDIR/include and $QTDIR/lib
will be used instead of calling qmake.

For Windows, it is necessary to set the environment variable SDLDIR to the
location where SDL is installed, and to set OSTYPE=windows.

  make

There shouldn't be any serious warnings from the compiler. Use "make
mklynxcc" if you want to also build a copy of mklynxcc (see below).

For Windows, mingw32-make must be used and the default Command Prompt (cmd)
is assumed, not the MSYS shell.

  make install

Running "make install" as root will do the following:

* Copy the tworld2 binary to /usr/local/games.
* Copy the tworld2.6 manpage to /usr/local/man/man6.
* Create /usr/local/share/tworld if it does not exist.
* Copy the external resources (i.e., the bitmaps and wave files) to
  /usr/local/share/tworld/res.
* Create the directories /usr/local/share/tworld/data and
  /usr/local/share/tworld/sets.
  
Running "make install2" instead will copy only the binary and manpage,
which is sufficient for upgrading existing installations of Tile World.

The sets directory is where you will generally store the .dat files that
you want to use. However, if you want to make use of a configuration file
with a particular data file, then you will need to store the data file in
the data directory, and the configuration file goes into the sets directory
instead. See the documentation for more information.

  Level Sets

As mentioned above, the original "Chip's Challenge" level set does not come
with Tile World, for reasons of copyright. If you do not already have a
copy of Microsoft's Windows version of "Chip's Challenge", you might still
be able to find a copy. Search the links listed below, under "Resources on
the Internet", for helpful hints on finding the game online.

If and when you do, you can copy the chips.dat file from there into Tile
World's data directory. You will then be able to play the levels of the
original set (both in MS mode and in Lynx mode).

There are also many "user-created" level sets. These are sets of levels
which have been invented by fans of the game. These sets are freely
available for downloading. If you have a .dat file that contains a level
set and you wish to use it, just copy it to Tile World's sets directory.
The next time you start Tile World, the new .dat should appear in the list
of available level sets.

At http://www.pillowpc2001.net/levels/ is a list of the available
level sets that have been created. I have not included any of these
level sets in this distribution, as the authors continue to add new
levels to their sets over time.

Actually, this distribution does contain one small level set. This is
included so that even if you don't have the original level set, you can
still get a brief glimpse of how the game works, and what some of the most
basic challenges are. Also, the same set of levels can be played with both
the MS and Lynx ruleset, so you can see how they differ.

Finally, there are some special level sets that deserve particular
mention.  They are CCLP2.dat and CCLP3.dat "Chip's Challenge Level
Packs 2 and 3".  These sets were assembled by the fans, who voted on
which levels to include. The levels that were voted as being the most
fun were then all put together. They are the closest things we have to
sequels for the original game.  (But be careful: they are both much
harder than the original!)

Also, there is CCLP1.dat. This set is especially intended for people
new to the game, conceived of as a replacement of the original
chips.dat for those who do not have it.

You can download a copy of Tile World with these sets already
installed. If instead you download them separately, you will want to
store them in your data directory instead of your sets
directory. (This is because Tile World comes with special
configuration files for CCLP1, CCLP2, and CCLP3.)  Note that only some
levels in CCLP2 can be played under the Lynx ruleset.  CCLP1 and
CCLP3, on the other hand, had compatibility with both rulesets as a
criterion, and hence all of its levels can be enjoyed in the ruleset
of your choice.

  The Complete Documentation

The full documentation for Tile World is included with the distribution, in
the file tworld2.html. There you will find information on how to play the
game, adding new level sets, customizing Tile World, and more.

  Creating New Level Sets

The most widely used program for creating new level sets is ChipEdit. It
comes with excellent documentation, and you should have little trouble
learning how to use it. Some other editors have recently been made
available, such as CCEdit and Chip's Workshop.

Normally, ChipEdit creates levels for the MS ruleset. If you wish to make a
level set for the Lynx ruleset, you have a few options:

* A very simple command-line utility is included with Tile World, called
  mklynxcc. This program will change a normal .dat file to one that will
  use the Lynx ruleset instead of the MS ruleset. Running mklynxcc foo.dat
  will change foo.dat's ruleset from MS to Lynx.
* You can use a configuration file to override the builtin ruleset. This
  method requires creating an extra file, but does allows you to avoid
  making changes to the .dat file. See the complete documentation for
  information on how to set up a configuration file.
* Finally, ChipEdit has an obscure feature which allows you to control the
  signature of the data file. This is done by adding a SIGNATURE entry to
  the chipedit.ini file. The default signature value is 0x0002AAAC, which
  indicates a data file that uses the MS ruleset. If you set the SIGNATURE
  to be 0x0102AAAC, then ChipEdit will create data files marked to use the
  Lynx ruleset instead.

  Resources on the Internet

There is quite a bit of information about "Chip's Challenge" available on
the internet. Much of it is focused on maximizing your score on the
original level set for the MS game, but you will also find lots of general
help and useful information as well.

Jimmy Vermeer maintains a site that tracks people's scores on the original
level set and CCLP2, as well as information on other levels and links to
many other pages. (But it hasn't been updated in a long time.)

  http://www.geocities.com/purpletentacle1977ca/

Mike L's Chip's Challenge hosts CCLP2 and CCLP3 and has links to almost
every custom level set available.

  http://www.pillowpc2001.net/

Anders Kaseorg's site contains the Chip's Challenge FAQ, as well as the AVI
repository and a web interface to the newsgroup:

  http://chips.kaseorg.com/

Many (but not all) user-created level sets can be found at the
chips_challenge Yahoo group:

  http://groups.yahoo.com/group/chips_challenge/files/

ChipEdit's home page is at:

  http://www.stage62.com/chipedit/chipedit.htm

Finally, the original Tile World home page is at:

  http://www.muppetlabs.com/~breadbox/software/tworld/
  
This latest version - Tile World 2 - is being maintained at:

  http://madhavshanbhag.sitesled.com/Landlubber/TileWorld/index.html

  License

Tile World is copyright (C) 2001-2014 by Brian Raiter, Madhav
Shanbhag, and Eric Schmidt.  This program is free software; you can
redistribute it and/or modify it under the terms of the GNU General
Public License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.  This
program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License,
included in this distribution in the file COPYING, for more details.

  Bugs

Bug reports are always appreciated, and can be sent to Madhav Shanbhag
at CrapulentCretin@Yahoo.com, or to Eric Schmidt at
eric41293@comcast.net. The list of known bugs is at
http://www.muppetlabs.com/~breadbox/software/tworld/BUGS.html. Please
check here before sending a bug report, to to make sure the bug has
not already been documented.

  Credits

Tile World was written by Brian Raiter.

Tile World 2 - which simply improves upon the user interface of the original
Tile World program - was developed by Madhav Shanbhag.

Version 2.1 was produced by Eric Schmidt.

The sound effects included in this distribution were created by Brian
Raiter, with assistance from SoX. Brian Raiter has explictly placed these
files in the public domain.

The tile images included in this distribution were created by Anders
Kaseorg, with assistance from POV-Ray. Anders Kaseorg has explicitly placed
these files in the public domain.

The introductory set of levels included in this distribution were created
by Brian Raiter. Brian Raiter has explictly placed these levels in the
public domain.

Thomas Harte and Michael Hansen have developed Mac OSX ports of Tile World and
Tile World 2 respectively.

Thanks to "The Architect" for his corrections to the font bitmap file.

"Chip's Challenge" was designed by Chuck Sommerville, who is also the
author of the original Lynx program.

"Chip's Challenge" is a registered trademark of Alpha Omega Publications.

Creating this program would have been flatly impossible without the help of
several fans of "Chip's Challenge". The author would particularly like to
acknowledge Anders Kaseorg for sharing the fruits of his investigations
into the game logic of the MS version and for being an effective bug
hunter, Chuck Sommerville for his pointers regarding the game logic of the
Lynx version and his unfailing support of this project, and "CCExplore" for
his in-depth investigations of esoteric game behavior.

Many other regulars of the annexcafe.chips.challenge newsgroup assisted
with bug reports, suggestions, and all-around encouragement. Their help is
gratefully acknowledged.

The anonymous author of the document describing the .dat file format, Don
Gregory, the "Charter Chipsters", and the contributors to the CC AVI
library all deserve mention as well -- this program would never have been
written without the information they made freely available.

Last but not least, a tip of the hat to John K. Elion for writing ChipEdit.


Changelog
2014 Mar 26: version 2.1.0 (by Eric Schmidt, with contributions from
             Madhav Shanbhag)
* Main window layout changed; in particular, passwords are visible.
* Fixed Lynx toggle wall bug: if a level stopped running the instant a
  toggle button was pressed, the walls would toggle the next time play
  started.
* When selecting a level set, the game now tries to start on the level the
  player left off from, rather than the earliest unsolved level.
* MS mouse emulation bug fixed. (report and fix from David Stolp)
* Some "stuck key" problems fixed, such as Alt-Tab on Windows 7.
* The fileio.c patch, which fixes a .tws corruption problem on 64-bit Linux,
  has been applied.
* Added the ability to disable the display of CCX text when visiting or
  completing a level.
* Added ability to view CCX text via menu options.
* Messages displayed when winning, dying, or running out of time are now
  customizable via external file.
* When playback of the solution to the last level in the set finishes, the
  congratulatory message box wouldn't appear. Now it does.
* In Lynx mode, it is now possible to select the starting direction for the
  "random" force floors, by using the "f" key.
* Added settings file. There are currently two settings: (i) whether to
  display CCX text, and (ii) sound volume.
* Debug/cheat keys completely disabled for releases.
* In pedantic Lynx mode, when an entity attempts to leave the map, rather
  than play being halted, a warning is printed on stderr, and play
  continues as in non-pedantic mode.
* "Goto" menu item renamed "Enter Password", and greyed out during play.
  (Before, it could be selected, but it would not do anything.)
* Miscellaneous code changes not visible to users, to eliminate compiler
  warnings and for other reasons.

2010 Dec 22: version 2.0.0 (by Madhav Shanbhag)
* First public release of Tile World 2 featuring numerous improvements.

2006 Apr 17: version 1.3.0
* Added mouse handling to the code. That includes modifying the SDL code to
  handle mouse activity, adding a large set of new commands, making it
  possible for Chip to move towards a goal over several ticks, and
  modifying the action struct (and the solution file) to permit storing the
  new moves in solutions.
* Added the database of unsolvable levels, which warns users before they
  play a level known to be unsolvable, and automatically gives them the
  password to the next level.
* Added the solution file display, allowing the user to manage multiple
  solution files for a single level set.
* Changed code to accept a solution file as the sole cmdline argument, if
  the solution file contains the set name. Hopefully this, and the previous
  addition, will make it easier for people to share solution files.
* Added the batch-mode verification feature, to check the existing
  solutions without initializing the GUI subsystem.
* Fixed the old emulation bug (MS logic) involving Chip turning a cloned
  block into a clone of himself while going through a teleport.
* Fixed the bug (MS logic) preventing Chip from pushing a slipping block
  parallel to its slipping direction when the push happens through a
  teleport.
* Improved the emulation of several creatures teleporting simultaneously
  (Lynx logic), with assistance from ccexplore.
* Minor improvements to the Lynx pedantic mode.
* Fixed the password input routine. (It was completely broken; not sure
  when that happened.)
* Rewrote the code so that level data and solutions are kept compressed in
  memory, and expanded only when needed, instead of decoding entire files
  when read. This change makes for a small reduction in memory usage, and
  reduces the overall level of useless work being done.
* Added code to shutter the map view when MS gameplay is paused.
* Numerous minor tweaks to the code, the display, and the documentation.

2006 Feb 24: version 1.2.2
* Finally got the handling of brown buttons (MS logic) working correctly.
  It's been a long time coming, and it's all so simple in hindsight.
* Added code to grok the new solution file format. This version of Tile
  World will continue to use the old solution file format, but if it
  encounters a file in the new format, it will handle it correctly.
* Added the ability to specify a solution file by name on the command-line.
* Added -P option for Lynx pedantic mode.
* Added the Shift-Ctrl-X feature to permanently delete solutions.
* Fixed bugs in Shift-Tab feature, and formally documented this feature.
* Fixed bug in the stepping UI.
* Fixed bug reported by Thomas Harte, where an international keyboard
  generating a non-Latin-1 character could cause TW to segfault.
* Fixed bug reported by Catatonic Porpoise, where SDL was being asked to
  blit to a locked surface.
* Fixed a segfault in MS logic reported by Evan Dummit (occurring when a
  cloner's creature in the northwest corner is replaced by Chip).
* Added casts to avoid new warnings in gcc 4.
* Fixed a few other minor MS emulation bugs reported by Evan Dummit.
* Added a number of unfixed MS emulation bugs to msbugs.txt, several of
  which were reported by Evan Dummit.
* A handful of documentation errors were also found and fixed.

2004 Oct 21: version 1.2.1
* Introduced a basic stepping-control UI, and added stepping information to
  the solution file format.
* Fixed a bug (MS logic) reported by David Stolp, involving blocks going
  out and back into the slip list so quickly that they weren't sent to the
  end of the list.
* Fixed a bug (MS logic) reported by David Stolp where block and creatures
  were not moving into key tiles if the bottom tile prevented it (e.g. a
  wall).
* Fixed a bug in sdltext.c, reported by Dale Bryan caused a segfault in
  16-bit color mode.
* Added -r (read-only) cmdline option.

2004 Oct 04: version 1.2.0
* Fixed a bug (MS logic) preventing cloned tanks from turning around when
  they hadn't yet left the clone machine.
* Fixed another longstanding bug (MS logic) involving a tank indirectly

defs.h
/* defs.h: Definitions used throughout the program.
 *
 * Copyright (C) 2001-2010 by Brian Raiter and Madhav Shanbhag,
 * under the GNU General Public License. No warranty. See COPYING for details.
 */

#ifndef	HEADER_defs_h_
#define	HEADER_defs_h_

#include	<stdio.h>
#include	<time.h>
#include	"gen.h"

/*
 * Miscellaneous definitions.
 */

/* The various rulesets the program can emulate.
 */
enum {
    Ruleset_None = 0,
    Ruleset_Lynx = 1,
    Ruleset_MS = 2,
    Ruleset_Count
};

/* File I/O structure.
 */
typedef	struct fileinfo {
    char       *name;		/* the name of the file */
    FILE       *fp;		/* the real file handle */
    char	alloc;		/* TRUE if name was allocated internally */
} fileinfo;

/* Pseudorandom number generators.
 */
typedef	struct prng {
    unsigned long	initial;	/* initial seed value */
    unsigned long	value;		/* latest random value */
    char		shared;		/* FALSE if independent sequence */
} prng;

/*
 * Definitions used in game play.
 */

/* A move is specified by its direction and when it takes place.
 */
typedef	struct action { unsigned int when:23, dir:9; } action;

/* A structure for managing the memory holding the moves of a game.
 */
typedef struct actlist {
    int			allocated;	/* number of elements allocated */
    int			count;		/* size of the actual array */
    action	       *list;		/* the array */
} actlist;

/* The range of relative mouse moves is a 19x19 square around Chip.
 * (Mouse moves are stored as a relative offset in order to fit all
 * possible moves in nine bits.)
 */
#define	MOUSERANGEMIN	(-9)
#define	MOUSERANGEMAX	(+9)
#define	MOUSERANGE	19

/* The complete list of commands that the user can given.
 */
enum {
    CmdNone = NIL,
    CmdNorth = NORTH,
    CmdWest = WEST,
    CmdSouth = SOUTH,
    CmdEast = EAST,
    CmdKeyMoveFirst = NORTH,
    CmdKeyMoveLast = NORTH | WEST | SOUTH | EAST,
    CmdMouseMoveFirst,
    CmdMoveNop = CmdMouseMoveFirst - MOUSERANGEMIN * (MOUSERANGE + 1),
    CmdMouseMoveLast = CmdMouseMoveFirst + MOUSERANGE * MOUSERANGE - 1,
    CmdReservedFirst,
    CmdReservedLast = 511,
    CmdAbsMouseMoveFirst,
    CmdAbsMouseMoveLast = CmdAbsMouseMoveFirst + CXGRID * CYGRID - 1,
    CmdMoveFirst = CmdKeyMoveFirst,
    CmdMoveLast = CmdAbsMouseMoveLast,
    CmdPrevLevel,
    CmdNextLevel,
    CmdSameLevel,
    CmdQuitLevel,
    CmdGotoLevel,
    CmdPrev,
    CmdNext,
    CmdSame,
    CmdPrev10,
    CmdNext10,
    CmdPauseGame,
    CmdHelp,
    CmdPlayback,
    CmdCheckSolution,
    CmdReplSolution,
    CmdKillSolution,
    CmdSeeScores,
    CmdSeeSolutionFiles,
    CmdVolumeUp,
    CmdVolumeDown,
    CmdStepping,
    CmdSubStepping,
    CmdRandomFF,
    CmdProceed,
#ifndef NDEBUG
    CmdDebugCmd1,
    CmdDebugCmd2,
#endif
    CmdQuit,
    CmdPreserve,
    CmdSeek,
#ifndef NDEBUG
    CmdCheatNorth,
    CmdCheatWest,
    CmdCheatSouth,
    CmdCheatEast,
    CmdCheatHome,
    CmdCheatKeyRed,
    CmdCheatKeyBlue,
    CmdCheatKeyYellow,
    CmdCheatKeyGreen,
    CmdCheatBootsIce,
    CmdCheatBootsSlide,
    CmdCheatBootsFire,
    CmdCheatBootsWater,
    CmdCheatICChip,
#endif
    CmdCount
};

/* True if cmd is a simple directional command, i.e. a single
 * orthogonal or diagonal move (or CmdNone).
 */
#define	directionalcmd(cmd)	(((cmd) & ~CmdKeyMoveLast) == 0)

/* The list of available sound effects.
 */
#define	SND_CHIP_LOSES		0
#define	SND_CHIP_WINS		1
#define	SND_TIME_OUT		2
#define	SND_TIME_LOW		3
#define	SND_DEREZZ		4
#define	SND_CANT_MOVE		5
#define	SND_IC_COLLECTED	6
#define	SND_ITEM_COLLECTED	7
#define	SND_BOOTS_STOLEN	8
#define	SND_TELEPORTING		9
#define	SND_DOOR_OPENED		10
#define	SND_SOCKET_OPENED	11
#define	SND_BUTTON_PUSHED	12
#define	SND_TILE_EMPTIED	13
#define	SND_WALL_CREATED	14
#define	SND_TRAP_ENTERED	15
#define	SND_BOMB_EXPLODES	16
#define	SND_WATER_SPLASH	17

#define	SND_ONESHOT_COUNT	18

#define	SND_BLOCK_MOVING	18
#define	SND_SKATING_FORWARD	19
#define	SND_SKATING_TURN	20
#define	SND_SLIDING		21
#define	SND_SLIDEWALKING	22
#define	SND_ICEWALKING		23
#define	SND_WATERWALKING	24
#define	SND_FIREWALKING		25

#define	SND_COUNT		26

/*
 * Structures for defining the games proper.
 */

/* The collection of data maintained for each level.
 */
typedef	struct gamesetup {
    int			number;		/* numerical ID of the level */
    int			time;		/* no. of seconds allotted */
    int			besttime;	/* time (in ticks) of best solution */
    int			sgflags;	/* saved-game flags (see below) */
    int			levelsize;	/* size of the level data */
    int			solutionsize;	/* size of the saved solution data */
    unsigned char      *leveldata;	/* the data defining the level */
    unsigned char      *solutiondata;	/* the player's best solution so far */
    unsigned long	levelhash;	/* the level data's hash value */
    char const	       *unsolvable;	/* why level is unsolvable, or NULL */
    char		name[256];	/* name of the level */
    char		passwd[256];	/* the level's password */
} gamesetup;

/* Flags associated with a saved game.
 */
#define	SGF_HASPASSWD		0x0001	/* player knows the level's password */
#define	SGF_REPLACEABLE		0x0002	/* solution is marked as replaceable */
#define	SGF_SETNAME		0x0004	/* internal to solution.c */

/* The history for the last time a levelset was played.
 */
typedef	struct history {
    char		name[256];	/* the set filename minus any path */
    char		passwd[256];	/* password of the last played level */
    int	   		levelnumber;	/* number of the last played level */
    struct tm		dt;		/* date/time set was last played */
} history;

/* The collection of data maintained for each series.
 */
typedef	struct gameseries {
    int			count;		/* number of levels in the series */
    int			allocated;	/* number of elements allocated */
    int			final;		/* number of the ending level */
    int			ruleset;	/* the ruleset for the game file */
    int			gsflags;	/* series flags (see below) */
    gamesetup	       *games;		/* the array of levels */
    fileinfo		mapfile;	/* the file containing the levels */
    char	       *mapfilename;	/* the name of said file */
    fileinfo		savefile;	/* the file holding the solutions */
    char	       *savefilename;	/* non-default name for said file */
    int			solheaderflags;	/* solution flags (none defined yet) */
    int			solheadersize;	/* size of extra solution header */
    char		filebase[256];	/* the level set's filename */
    char		name[256];	/* the filename minus any path */
    unsigned char	solheader[256];	/* extra solution header bytes */
} gameseries;

/* Flags associated with a series.
 */
#define	GSF_ALLMAPSREAD		0x0001	/* finished reading the data file */
#define	GSF_NOSAVING		0x0002	/* treat solution file as read-only */
#define	GSF_NODEFAULTSAVE	0x0004	/* don't use default tws filename */
#define	GSF_IGNOREPASSWDS	0x0008	/* don't require passwords */
#define	GSF_LYNXFIXES		0x0010	/* change MS data into Lynx levels */

#endif
